File: python/open_guardrail/guards/prototype_pollution.py
"""Detect prototype pollution attempts."""

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"__proto__"),
    re.compile(r"constructor\s*\.\s*prototype"),
    re.compile(r"constructor\s*\["),
    re.compile(r"Object\s*\.\s*assign\s*\("),
    re.compile(r"Object\s*\.\s*defineProperty\s*\("),
    re.compile(r'\["__proto__"\]'),
    re.compile(r"\['__proto__'\]"),
    re.compile(r"JSON\s*\.\s*parse\s*\(.*__proto__"),
    re.compile(
        r"prototype\s*\.\s*"
        r"(?:toString|valueOf|hasOwnProperty)\s*="
    ),
    re.compile(r'\{\s*"__proto__"\s*:'),
]


class _PrototypePollution:
    def __init__(
        self, *, action: str = "block"
    ) -> None:
        self.name = "prototype-pollution"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0) if triggered else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="prototype-pollution",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Prototype pollution attempt detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def prototype_pollution(
    *, action: str = "block"
) -> _PrototypePollution:
    return _PrototypePollution(action=action)

File: packages/guards/src/prototype-pollution.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface PrototypePollutionOptions {
  action: 'block' | 'warn';
}

const POLLUTION_PATTERNS: RegExp[] = [
  /__proto__/,
  /constructor\s*\.\s*prototype/,
  /constructor\s*\[/,
  /Object\s*\.\s*assign\s*\(/,
  /Object\s*\.\s*defineProperty\s*\(/,
  /\["__proto__"\]/,
  /\['__proto__'\]/,
  /JSON\s*\.\s*parse\s*\(.*__proto__/,
  /prototype\s*\.\s*(?:toString|valueOf|hasOwnProperty)\s*=/,
  /\{\s*"__proto__"\s*:/,
];

export function prototypePollution(options: PrototypePollutionOptions): Guard {
  return {
    name: 'prototype-pollution',
    version: '0.1.0',
    description: 'Detects prototype pollution attempts',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of POLLUTION_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'prototype-pollution',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
